### DISCUSSION TITLE
PHP intelephense lsp setting not working!

### DISCUSSION THREAD

**Discussion by Ok9xNirab** [zed-industries/zed#23788] [Feature Requests]
According to [intelephense  docs](https://github.com/bmewburn/intelephense-docs/blob/master/gettingStarted.md#environment), i want to set `wordpress` stub inside `intelephense.stubs`

So, i just configure the editor based on zed docs : 

```json
  "lsp": {
    "intelephense": {
      "initialization_options": {
        "stubs": [
          "apache",
          "bcmath",
          "bz2",
          "calendar",
          "com_dotnet",
          "Core",
          "ctype",
          "curl",
          "date",
          "dba",
          "dom",
          "enchant",
          "exif",
          "FFI",
          "fileinfo",
          "filter",
          "fpm",
          "ftp",
          "gd",
          "gettext",
          "gmp",
          "hash",
          "iconv",
          "imap",
          "intl",
          "json",
          "ldap",
          "libxml",
          "mbstring",
          "meta",
          "mysqli",
          "oci8",
          "odbc",
          "openssl",
          "pcntl",
          "pcre",
          "PDO",
          "pdo_ibm",
          "pdo_mysql",
          "pdo_pgsql",
          "pdo_sqlite",
          "pgsql",
          "Phar",
          "posix",
          "pspell",
          "random",
          "readline",
          "Reflection",
          "session",
          "shmop",
          "SimpleXML",
          "snmp",
          "soap",
          "sockets",
          "sodium",
          "SPL",
          "sqlite3",
          "standard",
          "superglobals",
          "sysvmsg",
          "sysvsem",
          "sysvshm",
          "tidy",
          "tokenizer",
          "xml",
          "xmlreader",
          "xmlrpc",
          "xmlwriter",
          "xsl",
          "Zend OPcache",
          "zip",
          "zlib",
          "wordpress"
        ]
      }
    }
  }
```

But didn't working !!

<img width="1552" alt="Screenshot 2024-01-21 at 18 53 19" src="https://github.com/zed-industries/community/assets/40653908/9e60b9ba-fc48-4000-a023-7e6432535044">

**Thanks**

**Comment by lukeholder** [comment]
I would love good PHP support in Zed. People don't realise how much PHP is used.

**Comment by builtbybasit** [comment]
I need support for PHP in Zed

**Comment by Tsolmonx** [comment]
Does anyone know zed supports intelephense premium ?

**Comment by tjohnman** [comment]
This is important for me, too. Is it a bug, or are we supposed to set the configuration in some other way?

**Comment by sbusch** [comment]
Premium should work, although not tested by myself: https://github.com/zed-industries/zed/issues/6965

**Comment by dinhtungdu** [comment]
I have a similar issue, `includePaths` is also not working for me. Here is the config I'm using:

```
  "lsp": {
    "intelephense": {
      "initialization_options": {
        "environment": {
          "includePaths": ["/path/to/lib"]
        }
      }
    }
  }
```

**Comment by ggarciajr** [comment]
braces format is also not being respected

**Comment by usizu** [comment]
Has anyone gotten Wordpress (and WooCommerce / ACF) stubs working in Zed?

**Comment by JeremyOB** [comment]
1+

**Comment by pjv** [comment]
This is a horrible kludge, but I got intelephense to respect the `includePaths` by directly editing one of the intelephense_data files in a private directory somewhere.

First I opened the zed log (`zed: open log`) and I scanned through until I saw something like this:

`2024-06-12T04:09:28-05:00 [INFO] open paths ["/private/var/folders/jk/h5pbftgn6nj_7w6rdqkqlg2w0000gq/T/intelephense/51d78cbd/intelephense_data_1"]`

That file looked like this:

`{"timestamp":1718183316791,"counter":{"next":1296,"holes":[]},"environment":{"documentRoot":"file:///<my wp project>“,"includePaths":[],"phpVersion":"8.3.0","shortOpenTag":false},"intelephenseVersion":"1.10.4"}`

I added a couple paths to `includePaths`:

`{"timestamp":1718183316791,"counter":{"next":1296,"holes":[]},"environment":{"documentRoot":"file:///<my wp project>“,"includePaths":["/path/to/wordpress","/path/to/woocommerce"],"phpVersion":"8.3.0","shortOpenTag":false},"intelephenseVersion":"1.10.4"}`

…then restarted zed and opened some php files with a bunch of wordpress function calls and no error messages from intelephense. Go to definition does not work, but mousing over wp function calls does show me the function’s template.

obviously there is some kind of disconnect between defining `lsp.intelephense.initialization_options` in your zed settings and those initialization_options getting propagated to intelephense.

**Comment by plugd-in** [comment]
I got LSP settings working for PHP, since I needed the stub for ssh2. Basically, I had to update the `zed_extension_api` for PHP to 0.0.6, and then add  a `fn language_server_workspace_configuration(...)` implementation which I basically grabbed from one of the other extensions. I had to build the extension using `zed/crates/extension_cli`, then I bumped the PHP extension version I built so it wouldn't be replaced, and copied the built extension to `.local/share/zed/extensions/installed`, I believe. I forked Zed and added my commit to my fork. After updating my PHP extension, I simply specified the stubs like so:
![image](https://github.com/zed-industries/zed/assets/84760193/7770f2cd-9fe6-4eab-b3f5-a2d1ce040dae)

If anyone wants to take over from here, feel free to fork Zed, cherry-pick my commit and then create a PR.

**Comment by federicoangelli** [comment]
> I got LSP settings working for PHP, since I needed the stub for ssh2. Basically, I had to update the `zed_extension_api` for PHP to 0.0.6, and then add a `fn language_server_workspace_configuration(...)` implementation which I basically grabbed from one of the other extensions. I had to build the extension using `zed/crates/extension_cli`, then I bumped the PHP extension version I built so it wouldn't be replaced, and copied the built extension to `.local/share/zed/extensions/installed`, I believe. I forked Zed and added my commit to my fork. After updating my PHP extension, I simply specified the stubs like so: ![image](https://private-user-images.githubusercontent.com/84760193/345958870-7770f2cd-9fe6-4eab-b3f5-a2d1ce040dae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA5NjcxMjcsIm5iZiI6MTcyMDk2NjgyNywicGF0aCI6Ii84NDc2MDE5My8zNDU5NTg4NzAtNzc3MGYyY2QtOWZlNi00ZWFiLWIzZjUtYTJkMWNlMDQwZGFlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE0VDE0MjAyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEzYzRiODUwNWM1NWNiNDcxZmMwZmE1NjAxNDEyYWViOTljYTBmMmU5NTEzYWU1OWFlYTlhNTY2ZTEyNWJhNWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.QvlRa1bjD9f_ySviHOKyeHQtAkTEjG_eay8h8-DH8Qo)
> 
> If anyone wants to take over from here, feel free to fork Zed, cherry-pick my commit and then create a PR.

Could you please give a more detailed step-by-step explanation of what you had to do? I've forked Zed, installed Rust and Xcode (per [these instructions](https://github.com/zed-industries/zed/blob/main/docs/src/development/macos.md)), modified the files as in your commit, ran `cargo build --release`, got that "target" directory, but I'm not sure what I should do next. I don't have the `.local/share/zed` directory. Also, can you build _just_ the extension? It seems that I have built the whole Zed app using that command. And how do I "bump" the extension's version? Thanks! 🙏

**Comment by plugd-in** [comment]
@federicoangelli Yeah, no problem. Basically, part of the release process, I assume is to build the extensions into a WASM plugin, and those plugins are downloaded from some plugin/release server. I think if you run Zed and install the PHP extension, you should have a `~/.local/share/zed/extensions/installed/php/` directory where the downloaded extension is placed.

Now, the tool that's actually used to build the extension (which was somewhat hard to figure out), is in `crates/extension_cli`. You can build it with `cargo build -p extension_cli --release`. After building the it, you can go on to build your extension like `mkdir /tmp/zed_build && mkdir extension_build && ./target/release/zed-extension --source-dir extensions/php --output-dir extension_build --scratch-dir /tmp/zed_build`. For details on the arguments, you can also use `./target/release/zed-extension --help`. You should end up with a file, `extension_build/archive.tar.gz`, with the extension in it. You can unpack it with `cd extension_build && tar -xzvf archive.tar.gz`, and your extension should look a lot like what was present in `~/.local/share/zed/extensions/installed/php/`.

You should then edit the manifest.json file to have some version greater than `0.0.2`, such as `0.0.3` or higher, which will prevent Zed from automatically replacing your extension with the version with lesser LSP setting support.

Then, you move your built extension into the `~/.local/share/zed/extensions/installed/php/` directory, replacing the files that were currently there. You can try to do so like: `cd .. && rm extension_build/archive.tar.gz && rm -rf ~/.local/share/zed/extensions/installed/php/ && cp -r extension_build ~/.local/share/zed/extensions/installed/php/`, which should move you out of the `extension_build` directory, remove the archive (you can also replace it with a `mv extension_build/archive.tar.gz .` if you want to keep it), remove the previously installed extension, and copy your built extension to the directory where extensions are installed. You'll also have to restart zed if you have it open.

Do note that I did all this in Linux, and the case might be different for Mac, although the biggest difference might be where the extensions are stored.

**Comment by federicoangelli** [comment]
@plugd-in You're right, I've managed to make it work following your steps. On macos, the path to where extensions are installed is `~/Library/Application\ Support/Zed/extensions/installed/`, that's the only difference in our context. Thanks! 🙏

**Comment by pjv** [comment]
@osiewicz Any chance of getting @plugd-in’s patch incorporated into the official PHP extension so I can avoid setting up a dev environment to be able to compile Zed? 😄

**Comment by maxdeviant** [comment]
v0.1.2 of the PHP extension now has support for reading the LSP settings from the Zed settings:

Here's how the settings should be structured in your `settings.json`:

```json
{
  "lsp": {
    "intelephense": {
      "settings": {
        "stubs": ["foo", "bar"],
        "files": {
          "maxSize": 1000001
        }
      }
    }
  }
}
```

**Comment by charmfocus** [comment]
> I have a similar issue, `includePaths` is also not working for me. Here is the config I'm using:
> 
> ```
>   "lsp": {
>     "intelephense": {
>       "initialization_options": {
>         "environment": {
>           "includePaths": ["/path/to/lib"]
>         }
>       }
>     }
>   }
> ```

https://github.com/zed-industries/zed/issues/16009#issuecomment-2278515131

**Comment by nadavmoreno** [comment]
this one is working for me:
```
  "lsp": {
    "intelephense": {
      "settings": {
        "environment": {
          "includePaths": ["/path/to/lib"]
        },
        "stubs": [
          "wordpress",
          "woocommerce",
          "acf-pro",
          "wordpress-globals",
          "wp-cli"
        ],
        "files": {
          "maxSize": 5000000,
          "excluded": ["**/node_modules/**", "**/vendor/**"]
        }
      }
    }
 }
```

**Comment by datgausaigon** [comment]
I have opened an issue:
https://github.com/zed-industries/zed/issues/17236
And I found this issue of yours related to it. Please join me in supporting it.
Thanks 
🐻

**Comment by admtech** [comment]
```
"lsp": {
    "intelephense": {
      "settings": {
        "environment": {
          "includePaths": [
            "/home/test/projects/framework",
            "/home/test/projects/core"
          ]
        },
        "stubs": ["openswoole"],
        "files": {
          "excluded": [
            "**/node_modules/**",
            "**/vendor/**",
            "**/old/**",
            "**/_old/**"
          ]
        }
      }
    }
  }

```
Unfortunately, this does **not work for me**. I just want to remove the directories /old/ and _old from code completion. But the setting is simply ignored and classes from _old are displayed. Also includePaths do not work.
My PHP extension v0.2.0 (with premium intelephense version) on linux (fedora) and macos (14.7)

Too bad, I would have loved to switch to the zed IDE for PHP development.

**Comment by crbanman** [comment]
For those who need to set up intelephense for Drupal, this is what works for me:

```
{
  "languages": {
    "PHP": {
      "language_servers": ["intelephense", "!phpactor"]
    }
  },
  "lsp": {
    "intelephense": {
      "settings": {
        "environment": {
          "includePaths": ["web/core/", "web/core/includes", "./vendor"],
          "phpVersion": "8.1.0",
          "documentRoot": "./web/index.php"
        },
        "files": {
          "associations": [
            "*.inc",
            "*.module",
            "*.install",
            "*.theme",
            "*.profile",
            "*.tpl.php",
            "*.test",
            "*.php",
            "*.info"
          ]
        },
        "format": {
          "braces": "k&r"
        }
      }
    }
  },
  "file_types": {
    "PHP": [
      "*.inc",
      "*.module",
      "*.install",
      "*.theme",
      "*.profile",
      "*.tpl.php",
      "*.test",
      "*.php",
      "*.info"
    ]
  }
}
```

**Comment by iansvo** [comment]
> v0.1.2 of the PHP extension now has support for reading the LSP settings from the Zed settings:
> 
> Here's how the settings should be structured in your `settings.json`:
> 
> {
>   "lsp": {
>     "intelephense": {
>       "settings": {
>         "stubs": ["foo", "bar"],
>         "files": {
>           "maxSize": 1000001
>         }
>       }
>     }
>   }
> }

This ended up being what I needed, thanks for sharing.

**Comment by notpeter** [comment]
I'm going to move this to the [Zed Language support discussions](https://github.com/zed-industries/zed/discussions/categories/language-support).

If folks have suggestions for incorporate additional configuration suggestions to always happy to PRs to [docs/languages/php.md](https://github.com/zed-industries/zed/blob/main/docs/src/languages/php.md?plain=1).

Thanks!

**Comment by andremacola** [comment]
Has anyone managed to make PHP_CodeSniffer (with a .phpcs.xml) work with Zed? Using Intelephense and phpactor at the same time is impossible and I prefer Intelephense.

Phpactor simply consumes 100% of my Mac M1 Pro every time I activate it and apparently there's some incompatibility with Intelephense.

I tried using an external formatter but without success.

`2025-02-26T02:50:41.73849-03:00 [ERROR] failed to format via external command "vendor/bin/phpcbf"`

```json
      "formatter": {
        "external": {
          "command": "vendor/bin/phpcbf",
          "arguments": [
            "-s",
            "--standard=.phpcs.xml",
            "{buffer_path}"
          ]
        }
      }
```

Without support for PHPCS, Zed becomes completely useless for those who work with PHP. Which is a shame.

**Reply by pjv** [reply]
this is working in my setup:

````
"PHP": {
      "language_servers": [
        "intelephense",
        "!phpactor",
        "!tailwindcss-language-server",
        "..."
      ],
      "format_on_save": "on",
      "formatter": {
        "external": {
          "command": "/usr/local/bin/bash",
          "arguments": [
            "-c",
            "cat > /tmp/zed_php_cs_fixer && php-cs-fixer fix --allow-risky=yes --config=/Users/pjv/.config/php-cs-fixer-wordpress/.php-cs-fixer.php --using-cache=no --quiet /tmp/zed_php_cs_fixer && cat /tmp/zed_php_cs_fixer"
          ]
        }
      },
````

i have no experience working with XML config files so i don't know what you might need to tweak, but since this is a completely external to zed command line call to the formatter i think you can do whatever you would natively do with phpcs.

**Comment by andremacola** [comment]
I don't use php-cs-fixer, I use phpcs/phpcbf, what I did was something similar but I created a shell script for that, since Zed does not allow saving a file in the formatter; the formatter needs to return to the buffer.

So I created an alias for the bash script named `zed-phpcbf`

```bash
#!/bin/bash
set -e

# Capture standard input
INPUT_CODE=$(cat)
if [ -z "$INPUT_CODE" ]; then
    echo "Error: No code provided via standard input." >&2
    exit 1
fi

# Check if PHPCBF is installed
PHPCBF_PATH=$(which phpcbf 2>/dev/null)
if [ ! -f "$PHPCBF_PATH" ]; then
    echo "Error: phpcbf not found. Check installation." >&2
    exit 1
fi

# Check if phpcs configuration file exists
PHPCS_CONFIG="./.phpcs.xml"
if [ ! -f "$PHPCS_CONFIG" ]; then
    echo "Error: Configuration file $PHPCS_CONFIG not found." >&2
    exit 1
fi

# Create temporary directory securely
TEMP_DIR=$(mktemp -d -t phpcbf-zed-XXXXXX) || { echo "Error: Failed to create temporary directory." >&2; exit 1; }
TEMP_FILE="$TEMP_DIR/file_$(date +%s)_$RANDOM.php"

# Cleanup function to ensure temporary files are removed
cleanup() {
    rm -rf "$TEMP_DIR"
}
trap cleanup EXIT

# Save code to a temporary file
echo "$INPUT_CODE" > "$TEMP_FILE" || { echo "Error: Failed to write to temporary file." >&2; exit 1; }

# Execute phpcbf with error handling
if ! phpcbf --standard="$PHPCS_CONFIG" "$TEMP_FILE" > /dev/null 2>&1; then
    # phpcbf returns non-zero exit status if it makes changes,
    # so we don't treat this as an error, just log it
    echo "Warning: phpcbf made changes to the code or encountered issues." >&2
fi

# Return the formatted code
if [ -f "$TEMP_FILE" ]; then
    cat "$TEMP_FILE"
else
    echo "Error: Temporary file was not found after processing." >&2
    exit 1
fi
```

**Zed Settings**

```json
    "PHP": {
      "language_servers": ["intelephense", "!phpactor"],
      "format_on_save": "on",
      "formatter": {
        "external": {
          "command": "phpcbf-zed"
        }
      }
    }
```

Unfortunately, Zed does not have a lint for phpcs.
Sublime, for example, has Sublimelinter with several addons for each need.
